### PR TITLE
task 1-5 implementation

### PR DESCRIPTION
**Pull request by olegasko**

### CODE REVIEW

File: src/main/java/com/expertsoft/tasks/OrderStats.java [hunk 1]
```diff
@@ -1,120 +1,148 @@
...
+     * <p>
      * Given a stream of orders, return a map, where keys are different order sizes and values are lists of orders,
      * referring to this sizes. Order size here is just a total number of products in the order.
      *
      * @param orders stream of orders
      * @return map, where order size values mapped to lists of orders
      */
     static Map<Integer, List<Order>> orderSizes(final Stream<Order> orders) {
-        return null;
+        Collection<Order> orderCollection = orders.collect(Collectors.toList());
```
Comment (KremnyovVlad) [inline src/main/java/com/expertsoft/tasks/OrderStats.java]: not really sure why do we need this line if in the next line we create stream again

File: src/main/java/com/expertsoft/tasks/OrderStats.java [hunk 2]
```diff
@@ -1,120 +1,148 @@
...
      * one product of the provided color and false otherwise.
      *
      * @param orders stream of orders
-     * @param color product color to test
+     * @param color  product color to test
      * @return boolean, representing if every order in the stream contains product of specified color
      */
     static Boolean hasColorProduct(final Stream<Order> orders, final Product.Color color) {
-        return null;
+        return orders.allMatch(order -> order.getOrderItems().stream().anyMatch(orderItem -> orderItem.getProduct().getColor().equals(color)));
```
Comment (KremnyovVlad) [inline src/main/java/com/expertsoft/tasks/OrderStats.java:67]: this method should check that all items of orders contain at least one product of the provided color
anyMatch method will return true even if just one order in  stream contains product of this color and this is not correct for current task

File: src/main/java/com/expertsoft/tasks/OrderStats.java [hunk 3]
```diff
@@ -1,120 +1,148 @@
...
      * @param cardNumber card number to check
      * @return average price of the product, ordered with the provided card
      */
     static BigDecimal averageProductPriceForCreditCard(final Stream<Customer> customers, final String cardNumber) {
         final AveragingBigDecimalCollector collector = new AveragingBigDecimalCollector();
-        return null;
+        Stream<BigDecimal> prices = customers.flatMap(customer -> customer.getOrders().stream())
+                .filter(order -> order.getPaymentInfo().getCardNumber().equals(cardNumber))
+                .flatMap(order -> order.getOrderItems().stream())
+                .flatMap(orderItem -> Collections.nCopies(orderItem.getQuantity(), orderItem.getProduct()).stream())
```
Comment (KremnyovVlad) [inline src/main/java/com/expertsoft/tasks/OrderStats.java:143]: not sure why do we need list of copies with size = quantity

Comment (olegasko) [reply src/main/java/com/expertsoft/tasks/OrderStats.java:143]: due to this comment in requirements:
--------------------------------------------------------------
     * [
     * Product1(price = 100$, quantity = 2),
     * Product2(price = 160$, quantity = 1)
     * ]
     * then the average product price for this order will be 120$ ((100 * 2 + 160 * 1) / 3)
-------------------------------------------------------------
we need to multiply items on quantity

File: src/main/java/com/expertsoft/tasks/OrderStats.java (hunk 1 above)
Comment (olegasko) [reply src/main/java/com/expertsoft/tasks/OrderStats.java]: left from debug, fixed

File: src/main/java/com/expertsoft/tasks/OrderStats.java (hunk 2 above)
Comment (olegasko) [reply src/main/java/com/expertsoft/tasks/OrderStats.java:67]: no, according to req: "every order contains at least one product" so, I've performed allMatch for items and anyMatch for products. I think this is proper implementation )

**Pull request closed without merging**